client/examples/perf.rs

```diff
@@ -0,0 +1,466 @@
+            let password = args.key_password.clone();
+            let secret = load_or_generate(&key_file, password);
+            ClientBuilder::from_url(address).crypto(FallbackCryptoProvider::new(secret))
+        } else {
+            ClientBuilder::from_url(address)
+        };
+
+        let client = builder.build().await?;
+        client.neighbours(0).await?;
+
+        Ok(client)
+    }
+
+    /// Establishes {connections} connections sequentially, appending to the {conns} buffer.
+    pub async fn establish_connections(
+        conns: &mut Vec<Client>,
+        args: &Options,
+        connections: usize,
+    ) -> anyhow::Result<()> {
+        for _ in 0..connections {
+            let client = establish_connection(args).await?;
+            conns.push(client);
+        }
+
+        Ok(())
+    }
+}
+
+mod connections_test {
+    use super::*;
+
+    #[derive(Debug)]
+    struct DataPoint {
+        connections: u32,
+        new_connections_rate: f32,
+    }
+
+    #[derive(Debug)]
+    struct Measurements {
+        data: Vec<DataPoint>,
+    }
+
+    impl Measurements {
+        fn new() -> Self {
+            Measurements { data: Vec::new() }
+        }
+
+        fn push(&mut self, connections: u32, new_connections_rate: f32) {
+            self.data.push(DataPoint {
+                connections,
+                new_connections_rate,
+            });
+        }
+
+        fn save(&self, writer: impl std::io::Write) -> anyhow::Result<()> {
+            let mut w = csv::Writer::from_writer(writer);
+
+            w.write_record(&["connections", "new-connections-rate"])?;
+            for data_point in &self.data {
+                w.write_record(&[
+                    &data_point.connections.to_string(),
+                    &data_point.new_connections_rate.to_string(),
+                ])?;
+            }
+
+            Ok(())
+        }
+    }
+
+    /// Establish connections in batches while asserting that none have been dropped
+    async fn measure_connections(
+        args: &Options,
+        conn_cmd: &ConnectionsCommand,
+    ) -> anyhow::Result<Measurements> {
+        let mut conns = Vec::new();
+        let connections = conn_cmd.connections_increment;
+
+        let mut measurements = Measurements::new();
+        'test_loop: loop {
+            let start = Instant::now();
+            let res = util::establish_connections(&mut conns, &args, connections).await;
+            if res.is_err() {
+                break 'test_loop;
+            }
+
+            let elapsed = Instant::now() - start;
+            let rate = (connections as f32) / elapsed.as_secs_f32();
+            log::info!(
+                "Connections currently established: {}, {:.1}/s",
+                conns.len(),
+                rate
+            );
+
+            for conn2 in conns.windows(2) {
+                let res = conn2[0].find_node(conn2[1].node_id()).await;
+                if res.is_err() {
+                    break 'test_loop;
```

Comment:
Are you sure that it is desired behavior to return Ok(measurements) even on error here and in establish connections? From user's perspective there will be only no print in line 179, nothing else I believe.

Reply:
Yes, both connection and load tests are expected to (usually) fail at some point due to limited ports or timeouts and we want performance data up to that point, AFAIK.